feat(templates): add amazonlinux-2023 distro

electricalgorithm opened the pull request:
This commit adds a simple template for the Amazon Linux 2023 operating system's latest version. The OS is scheduled to release every two weeks, therefore, we need to find a way to automatically update it from the following AWS website:

- https://cdn.amazonlinux.com/al2023/os-images/latest/

electricalgorithm committed:
feat(templates): add amazonlinux-2023 distro

This commit adds a simple template for the Amazon Linux 2023 operating
system's 2026-04-06 version. The OS is scheduled to release every two weeks,
therefore, a new script called hack/update-template-amazonlinux.sh
is added next to the other template updater scripts.

Find the latest images here:
https://cdn.amazonlinux.com/al2023/os-images/latest/

This patch:
- Adds amazonlinux-2023 template with 2026-04-06 release.
- Resolves mount issues on AL2023 with host OS.
release.
- Adds update-template-amazonlinux.sh script to help maintainers
automatically update the verison.

Signed-off-by: Gyokhan Kochmarla <hello@gyokhan.com>

electricalgorithm commented:
Testing it with:
```
$ LIMA_TEMPLATES_PATH=lima/templates limactl start --name al2023-test --arch aarch64 --cpus 4 template:amazonlinux-2023 --debug
```

Correct os-release Info:
```
[gokhankocmarli@lima-al2023-test ~]$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2023"
ID="amzn"
ID_LIKE="fedora"
VERSION_ID="2023"
PLATFORM_ID="platform:al2023"
PRETTY_NAME="Amazon Linux 2023.9.20251208"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2023"
HOME_URL="https://aws.amazon.com/linux/amazon-linux-2023/"
DOCUMENTATION_URL="https://docs.aws.amazon.com/linux/"
SUPPORT_URL="https://aws.amazon.com/premiumsupport/"
BUG_REPORT_URL="https://github.com/amazonlinux/amazon-linux-2023"
VENDOR_NAME="AWS"
VENDOR_URL="https://aws.amazon.com/"
SUPPORT_END="2029-06-30"
```

Correct uname:
```
[gokhankocmarli@lima-al2023-test ~]$ uname -a
Linux lima-al2023-test 6.1.158-180.294.amzn2023.aarch64 #1 SMP Mon Dec  1 05:36:18 UTC 2025 aarch64 aarch64 aarch64 GNU/Linux
```

Correct repos:
```
[gokhankocmarli@lima-al2023-test ~]$ yum repolist
repo id                                                                         repo name
amazonlinux                                                                     Amazon Linux 2023 repository
kernel-livepatch                                                                Amazon Linux 2023 Kernel Livepatch repository
```

electricalgorithm commented:
It's weird, but the only problem is that the backspace does not work. I don't know if Lima uses SSH underneath, but I'll investigate the debug logs by re-shelling to the VM.

[Update] Yep, it looks like it uses SSH.
```
DEBU[0000] Failed to detect CPU features. Assuming that AES acceleration is available on this Apple silicon.
DEBU[0000] OpenSSH version 10.0.2 detected, is GSSAPI supported: true
DEBU[0000] AES accelerator seems available, prioritizing aes128-gcm@openssh.com and aes256-gcm@openssh.com
DEBU[0000] executing ssh (may take a long)): [/usr/bin/ssh -F /dev/null -o IdentityFile="/Users/gokhankocmarli/.lima/_config/user" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o NoHostAuthenticationForLocalhost=yes -o PreferredAuthentications=publickey -o Compression=no -o BatchMode=yes -o IdentitiesOnly=yes -o GSSAPIAuthentication=no -o Ciphers="^aes128-gcm@openssh.com,aes256-gcm@openssh.com" -o User=gokhankocmarli -o ControlMaster=auto -o ControlPath="/Users/gokhankocmarli/.lima/al2023-test/ssh.sock" -o ControlPersist=yes -t -o SendEnv=COLORTERM -o LogLevel=ERROR -p 54658 127.0.0.1 -- cd /Users/gokhankocmarli/ || cd /Users/gokhankocmarli ; exec "$SHELL" --login]
```

Then I believe it could be an issue with SendEnv.

afbjorklund commented:
* https://github.com/lima-vm/lima/issues/4047

Previous attempt with this (Amazon Linux) was:

* https://github.com/lima-vm/lima/pull/2537

I think it also failed to automate the updating?

electricalgorithm commented:
> * [Add amazon linux template #4047](https://github.com/lima-vm/lima/issues/4047)
> 
> 
> Previous attempt with this (Amazon Linux) was:
> 
>     * [Create amazonlinux2023 template #2537](https://github.com/lima-vm/lima/pull/2537)
> 
> 
> I think it also failed to automate the updating?

Thank you for the info! I have reviewed the comments there and see that two concerns are in common:

- No dynamic approach to retrieve the latest image from CDN. (Should we even want this?)
- Mount problems with virtiofs.

I'll spend some time fixing those. However, I believe that we can move this template to experimental ones and start iteratively on fixing stuff. Some folks may not need mounts or be happy with just having AL2023 with Lima without the latest image since it's possible to easily update the system using `dnf upgrade --releasever=<new-release-date>`. Or even better, use `latest` for releasever.

We could easily patch it by adding a run-script when the image is initialised and upgrading the system to the latest. Wdyt?

electricalgorithm commented:
**Update**: The mount problem is gone, and it's possible to update the image to the latest one within a new script under `hack/`. Additionally, the system tries to update the latest release when the machine is initialised automatically.

I'll check the unit test and the lining stuff later. Does it look OK to you?

AkihiroSuda commented on templates/experimental/amazonlinux-2023.yaml:
```diff
@@ -0,0 +1,58 @@
+
+mounts:
+- location: "~"
+  mountPoint: "/mnt/host"
```
The path is inconsistent with other template

AkihiroSuda commented on templates/experimental/amazonlinux-2023.yaml:
Should be symlinked as `amazonlinux.yaml`

AkihiroSuda commented on hack/update-template-amazonlinux-2023.sh:
```diff
@@ -0,0 +1,52 @@
+set -euo pipefail
+
+SETUP_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TEMPLATE_FILE="${SETUP_DIR}/../templates/experimental/amazonlinux-2023.yaml"
```
The template file must not be hard-coded.
See other updaters.

AkihiroSuda commented on hack/update-template-amazonlinux-2023.sh:
Should be called from https://github.com/lima-vm/lima/blob/master/hack/update-template.sh

AkihiroSuda commented on templates/experimental/amazonlinux-2023.yaml:
```diff
@@ -0,0 +1,58 @@
+minimumLimaVersion: 2.0.0
```
Probably this is going to be:

```suggestion
minimumLimaVersion: 2.1.0
```

electricalgorithm commented on hack/update-template-amazonlinux-2023.sh:
Done, thanks!

electricalgorithm commented on templates/amazonlinux-2023.yaml:
```diff
@@ -0,0 +1,58 @@
+  system: false
+  user: false
+
+provision:
```
I don't know what the contextual difference is between provision on the template level and common storage, but I will move the AL-specific connection patches there directly. The one with the mount fix should be in the template.

electricalgorithm replied:
Move all the weird patches to boot/. Thanks

afbjorklund commented:
There is `make lint`, but keeping it in sync with `.github` is a constant battle... Maybe https://github.com/nektos/act

electricalgorithm commented on templates/amazonlinux-2023.yaml:
```diff
@@ -0,0 +1,15 @@
+  script: |
+    #!/bin/bash
+    set -eux -o pipefail
+    # Update to the latest release on the first boot
```
It have a different effect then giving releasever. Amazon Linux uses a "fixed" repository method where every two week, they release another repo. This helps customers to make sure that whenever they use the AL2023 with releasever=20251208, all packages would be the same. So, if one wants to update their packages, they have to upgrade the release version of the repository. 

Standard yum/dnf upgrade or update is not going to work since it does not define which version to upgrade the repo.

electricalgorithm replied:
I understand your point. However, IMO, the updates can be automatically applied since:

- AL2023 have an compatibility promise -- where there cannot be any update that breaks the system packages' compatibility. Therefore, there should not be any concern from any user to use the latest.
- AL2023's package updates are mainly regarding to CVE fixes. As far as I know mainstream users, they should be able to always get the latest security fixes. In the end, it's the same AL2023, with better security.

So, that's why I'd like to stay with the current auto-update until there's an issue ticket. Until then, let's assume that our users wants to pull the latest AL2023 (imagine it like latest Ubuntu 24.04 -- and not 24.04.3 or 24.04.4 for example)  

Would you prefer I remove the provision script entirely and let users manually update if desired?

AkihiroSuda replied:
> Would you prefer I remove the provision script entirely and let users manually update if desired?

That seems more consistent with other templates

electricalgorithm replied:
OK. I removed the provision auto-update script from the template. Let's see if there will be people asking for it.

electricalgorithm commented on templates/_images/amazonlinux-2023.yaml:
```diff
@@ -0,0 +1,7 @@
+images:
+- location: "https://cdn.amazonlinux.com/al2023/os-images/2023.9.20251208.0/kvm/al2023-kvm-2023.9.20251208.0-kernel-6.1-x86_64.xfs.gpt.qcow2"
```
The URL is permenant. They will be all there. For any case, I added the hack script.

electricalgorithm commented:
> I still don't want new commits to Lima every two (just to bump a snapshot version) every two weeks.

It should not be. We can update it on every two weeks by the hack script, but in the end provision script will handle it for us automatically.

electricalgorithm commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,29 @@
+
+# Workaround for Amazon Linux 2023
+if [ -f /etc/os-release ] && grep -q "Amazon Linux" /etc/os-release; then
+	# 1. Create missing mount.virtiofs helper
```
I can cut a ticket to the team, but I think it should not be the concern of this PR. We can go with the current upstream approach, and when the fix is there, remove the patch here.

AkihiroSuda replied:
>  I think it should not be the concern of this PR

Why not.

electricalgorithm commented:
Hey people,

Sorry for being late. Quite busy months in Winter, as you can imagine. I updated the PR with a change to the latest image and addressed your comments. Please review the code once more, although it does not have much changes.

There seems to be issue on GitHub's CI checks. The Lints job failed due to a README's link issue which the PR does not have any modification. I believe there's another issue there. On the vz's integation test job, I couldn't understand if that was my changes breaking it, but after discussing with GitHub's Copilot, looks like it is regarding to setup of the job.

Can you verify that the failures are not related to the PR? If all looks good, please check the version string in the template -- which is 1.2.1 right now and check if it's possible to merge it until the release window closes.

Kindest regards,
Gyokhan 

cc: @afbjorklund @AkihiroSuda

AkihiroSuda commented on templates/amazonlinux.yaml:
```diff
@@ -0,0 +1 @@
+amazonlinux-2023.yaml
```
templates/README.md has to be updated too

electricalgorithm replied:
I'm updating the commit to add it to the README. Thanks!

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,29 @@
+		MOUNT_POINT=$(grep "mount0" "${USER_DATA}" | awk -F, '{print $2}' | tr -d '[:space:]')
+		if [ -n "${MOUNT_POINT}" ] && ! mountpoint -q "${MOUNT_POINT}"; then
+			mkdir -p "${MOUNT_POINT}"
+			mount -t virtiofs mount0 "${MOUNT_POINT}" || true
```
What about mount1, mount2, ... ?
Only mount0 fails?

electricalgorithm replied:
you're right. I'll update the commit to automatically read the user-data and add all the mounts not only mount0.

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,29 @@
+		chmod +x /sbin/mount.virtiofs
+	fi
+
+	# Cloud-init fails to mount 'mount0' because it doesn't recognize it as a device,
```
What is the cause of the issue?
Why does this issue only happen with AL?

electricalgorithm replied:
Although the kernel includes the support for virtiofs, there's no helper script under `/sbin`, therefore cloudinit fails. The first part of the patch creates the mount script and makes it executable.

```
[gokhankocmarli@lima-al2023-test ~]$ ls -la /sbin/mount.*
-rwxr-xr-x. 1 root root 200744 Jan 31  2023 /sbin/mount.fuse3
-rwsr-xr-x. 1 root root 219744 Feb  1  2023 /sbin/mount.nfs
lrwxrwxrwx. 1 root root      9 Feb  1  2023 /sbin/mount.nfs4 -> mount.nfs
```
```
[gokhankocmarli@lima-al2023-test ~]$ grep -i virtiofs /proc/filesystems
nodev   virtiofs
```

The second half of the script tries to mount all the mount points defined the user script if the cloud init failed due to missing helper script.


But, you're right. I'll update the commit to automatically read the user-data and add all the mounts not only mount0.

electricalgorithm commented:
Please review again :) I updated the code to apply your suggestions.

electricalgorithm commented:
Looks like the failure in jobs is due to the flakiness. Not related to the change.

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,38 @@
+	# Amazon Linux 2023 requires a workaround for virtiofs mounts.
+	# The mount.virtiofs helper script is missing, and cloud-init
+	# may fail to mount the filesystems. This creates the helper
+	# and manually mounts them.
```
Does it still produce the valid `/etc/fstab`? If so, just running `mount -a` might be fine here?

electricalgorithm replied:
I start a new AL2023 KVM to check and it seems that the fstab does not have the virtiofs configs. Therefore, mount -a does not work and we need to parse them from cloud init config's user-data. That was a nice catch by the way. I'm not super into OSes, thanks for the advice!

AkihiroSuda replied:
Why doesn't it generate fstab?

`/var/log/cloud-init.log`:
```
2026-04-16 07:52:01,442 - util.py[DEBUG]: Cloud-init v. 22.2.2 running 'init-local' at Thu, 16 Apr 2026 07:52:01 +0000. Up 1.95 seconds.
[...]
2026-04-16 07:52:04,763 - cc_mounts.py[DEBUG]: mounts configuration is [['lima-648190f135cc8172', '/Users/suda', 'virtiofs', 'ro,nofail', '0', '0']]
2026-04-16 07:52:04,763 - util.py[DEBUG]: Reading from /etc/fstab (quiet=False)
2026-04-16 07:52:04,763 - util.py[DEBUG]: Read 217 bytes from /etc/fstab
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Attempting to determine the real name of lima-648190f135cc8172
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: changed lima-648190f135cc8172 => None
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Ignoring nonexistent named mount lima-648190f135cc8172
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Attempting to determine the real name of ephemeral0
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: changed default device ephemeral0 => None
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Ignoring nonexistent default named mount ephemeral0
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Attempting to determine the real name of swap
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: changed default device swap => None
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Ignoring nonexistent default named mount swap
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: Skipping nonexistent device named None
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: no need to setup swap
2026-04-16 07:52:04,764 - cc_mounts.py[DEBUG]: No modifications to fstab needed
2026-04-16 07:52:04,764 - handlers.py[DEBUG]: finish: init-network/config-mounts: SUCCESS: config-mounts ran successfully
```

AkihiroSuda replied:
Maybe related to "AL2023 contains a customized version of cloud-init."

https://docs.aws.amazon.com/linux/al2023/ug/cloud-init.html

electricalgorithm commented:
Can you re-run the GitHub jobs and merge it if looks good to you since all conversations are resolved @AkihiroSuda?

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,38 @@
+	USER_DATA="/var/lib/cloud/instance/user-data.txt"
+	if [ -f "${USER_DATA}" ]; then
+		# Parse all mount entries from user-data (mount0, mount1, ...)
+		MOUNT_ENTRIES=$(grep -E '^\s*-\s+\[mount[0-9]+,' "${USER_DATA}" || true)
```
This grep didn't work for me

```console
[root@lima-amazonlinux ~]# head  /var/lib/cloud/instance/user-data.txt
#cloud-config
# vim:syntax=yaml

growpart:
  mode: auto
  devices: ['/']
mounts:
- [lima-648190f135cc8172, /Users/suda, virtiofs, "ro,nofail", "0", "0"]
timezone: Asia/Tokyo

[root@lima-amazonlinux ~]# grep -E '^\s*-\s+\[mount[0-9]+,' /var/lib/cloud/instance/user-data.txt
```

AkihiroSuda replied:
@electricalgorithm

How did you test your PR?
In what condition `user-data` can have mounts named `mount0`, `mount1`, etc.?

AkihiroSuda replied:
ping @electricalgorithm

AkihiroSuda commented on pkg/cidata/cidata.TEMPLATE.d/boot.Linux/04-amazonlinux-virtiofs.sh:
```diff
@@ -0,0 +1,38 @@
+
+# Workaround for Amazon Linux 2023
+if [ -f /etc/os-release ] && grep -q "Amazon Linux" /etc/os-release; then
+	# 1. Create missing mount.virtiofs helper
```
Why is this helper needed in the first place when we have to execute the mount command by ourselves?

AkihiroSuda replied:
The helper doesn't even seem to exist on other distros, such as almalinux-9
